mule-extensions-api-persistence/src/main/java/org/mule/runtime/extension/internal/persistence/OAuthGrantTypeTypeAdapter.java

```diff
@@ -0,0 +1,127 @@
+
+  private static final String GRANT_TYPE = "grantType";
+  private static final String AUTH_CODE = "Authorization Code";
+  private static final String CLIENT_CREDENTIALS = "Client Credentials";
+  private static final String TOKEN_URL = "tokenUrl";
+  private static final String EXPIRATION_REGEX = "expirationRegex";
+  private static final String DEFAULT_SCOPES = "defaultScopes";
+  private static final String CREDENTIALS_PLACEMENT = "credentialsPlacement";
+  private static final String ACCESS_TOKEN_EXPR = "accessTokenExpr";
+  private static final String REFRESH_TOKEN_EXPR = "refreshTokenExpr";
+  private static final String DEFAULT_SCOPE = "defaultScope";
+  private static final String ACCESS_TOKEN_URL = "accessTokenUrl";
+  private static final String AUTHORIZATION_URL = "authorizationUrl";
+
+  @Override
+  public void write(JsonWriter out, OAuthGrantType value) throws IOException {
+    try {
+      value.accept(new OAuthGrantTypeVisitor() {
+
+        @Override
+        public void visit(AuthorizationCodeGrantType grantType) {
+          try {
+            out.beginObject();
+            out.name(GRANT_TYPE).value(AUTH_CODE);
+            out.name(ACCESS_TOKEN_URL).value(grantType.getAccessTokenUrl());
+            out.name(AUTHORIZATION_URL).value(grantType.getAuthorizationUrl());
+            out.name(ACCESS_TOKEN_EXPR).value(grantType.getAccessTokenExpr());
+            out.name(EXPIRATION_REGEX).value(grantType.getExpirationRegex());
+            out.name(REFRESH_TOKEN_EXPR).value(grantType.getRefreshTokenExpr());
+            writeOptional(out, DEFAULT_SCOPE, grantType.getDefaultScope());
+            out.endObject();
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        }
+
+        @Override
+        public void visit(ClientCredentialsGrantType grantType) {
+          try {
+            out.beginObject();
+            out.name(GRANT_TYPE).value(CLIENT_CREDENTIALS);
+            out.name(TOKEN_URL).value(grantType.getTokenUrl());
+            out.name(ACCESS_TOKEN_URL).value(grantType.getAccessTokenExpr());
+            out.name(EXPIRATION_REGEX).value(grantType.getExpirationRegex());
+            writeOptional(out, DEFAULT_SCOPES, grantType.getDefaultScopes());
+            out.name(CREDENTIALS_PLACEMENT).value(grantType.getCredentialsPlacement().name());
+            out.endObject();
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        }
+      });
+    } catch (RuntimeException e) {
+
+      if (e.getCause() instanceof IOException) {
+        throw (IOException) e.getCause();
+      }
+
+      throw e;
+    }
+  }
+
+  @Override
+  public OAuthGrantType read(JsonReader in) throws IOException {
+    JsonObject json = new JsonParser().parse(in).getAsJsonObject();
+
+    // consider grant type could not be there for backwards compatibility
+    String grantType = json.has(GRANT_TYPE) ? json.get(GRANT_TYPE).getAsString() : null;
```

Comment:
why not set `AUTH_CODE` here directly instead of `null`?

Reply:
you could also call `getOptionalValue` which odes the same, allowing that method to receive the default value